src/test/java/com/woowacamp/soolsool/core/liquor/domain/LiquorCtrTest.java

```diff
@@ -35,12 +35,8 @@ void createFailWithOutNullLiquorId() {
     @DisplayName("클릭률을 구한다")
     void getCtr() {
         /* given */
-        LiquorCtr liquorCtr = LiquorCtr.builder()
-            .liquorId(1L)
-            .build();
-        liquorCtr.increaseImpressionOne();
-        liquorCtr.increaseImpressionOne();
-        liquorCtr.increaseClickOne();
+        LiquorCtr liquorCtr = new LiquorCtr(1L);
+        liquorCtr.overwrite(new LiquorCtr(1L, 2L, 1L));
```

Comment:
```suggestion
        final LiquorCtr liquorCtr = LiquorCtr.builder()
            .liquorId(1L)
            .impression(2L)
            .click(1L)
            .build();
```

뭔가 해당 given 절에서는 overwrite도 있어서 순수한 `getCtr()` 테스트가 아닌 것 같아요.

위처럼 변경하고 overwrite 테스트도 추가하면 어떨까요?

Reply:
첫 번째 코멘트 답변:
1. Redis 데이터가 만료됨 + 만료 이벤트 발생
2. 만료 이벤트를 통해 만료된 데이터가 WAS 적재됨
3. WAS 상에서 지연 쓰기 관련 만료 로직 처리 (RDB에 반영하는 로직)

요 세 단계는 이벤트 관련 로직이구요

코멘트 남겨주신 코드는 그냥 redis에 적재되어 있던 데이터를 가져오는 코드입니다!

두 번째 코멘트 답변:
크게 놓치고 있었는데 감사함다!!!!

Reply:
> 요 세 단계는 이벤트 관련 로직이구요
> 코멘트 남겨주신 코드는 그냥 redis에 적재되어 있던 데이터를 가져오는 코드입니다!

그럼 LiquorCtr 단위 테스트에서 만료 이벤트를 고려하고 있는건가요?

Reply:
아 1 ~ 4번 내용 코멘트가 월디가 아니라 와타사메 코멘트였군여!

계속 생각해 봤지만 사실 저는 NPE에 관련된 이야기를 잘 이해하지 못하겠어요..!

캐시(Redis)를 조회하려 했을 때,
1. miss -> RDB로부터 조회하여 캐시를 새로 등록한 뒤 해당 데이터를 반환한다.
2. hit -> 기존에 저장되어 있던 데이터를 반환한다.

lookUpLiquorCtr 메서드는 이러한 로직입니다!

> 1. Redis에 올라온 데이터를 WAS 메모리에서 가져옴
> 2. 가져오자마자 Redis에서 해당 데이터가 만료
> 3. 만료된 데이터가 WAS에 돌아다님
> 4. 따라서 Redis에서 데이터가 만료되었지만 WAS 메모리에는 남아있으므로 NPE 방지

일단 Redis에서 데이터를 가져오자마자 데이터가 만료되는 게 아닙니다.
정해진 TTL이 지나 해당 데이터가 만료(삭제)되면, Redis 만료 이벤트가 발생합니다.
이후 Redis 만료 이벤트 리스너가 해당 이벤트를 감지하여 WAS에서 지연 쓰기를 하는 이벤트를 발생시킵니다.
지연 쓰기가 완료되면 Redis에서 만료된 데이터는 남지 않고 그냥 사라집니다.

무엇보다 코멘트를 달아주신 코드를 보면 해당 코드는 그냥 캐시를 조회하는 메서드입니다. 단지 hit, miss 여부에 따라서 RDB에서 조회해올지 말지 등 처리가 달라지는 것뿐이에요.

Reply:
> 일단 Redis에서 데이터를 가져오자마자 데이터가 만료되는 게 아닙니다.

만료됐다는 상황을 가정한거였습니다! 하하 역시 소통은 어렵군요...

> 지연 쓰기가 완료되면 Redis에서 만료된 데이터는 남지 않고 그냥 사라집니다.

넵 근데 Redis에서 메모리로 가져온 데이터는 참조를 들고 있으면 언제든 재사용할 수 있는거 아닌가요? Redis에서 만료된 데이터가 WAS에서 계속 참조 하고 있으면 돌아다닌다는 것을 말하고 싶었습니다.

> 무엇보다 코멘트를 달아주신 코드를 보면 해당 코드는 그냥 캐시를 조회하는 메서드입니다. 

> 첫 번째 코멘트 답변:
> Redis 데이터가 만료됨 + 만료 이벤트 발생
> 만료 이벤트를 통해 만료된 데이터가 WAS 적재됨
> WAS 상에서 지연 쓰기 관련 만료 로직 처리 (RDB에 반영하는 로직)
> 요 세 단계는 이벤트 관련 로직이구요
> 코멘트 남겨주신 코드는 그냥 redis에 적재되어 있던 데이터를 가져오는 코드입니다!

`LiquorCtrTest`는 도메인 단위 테스트인데 Redis를 고려하는게 이상하다는 점에서 시작한 코멘트였습니다. 아무리 생각해도 `LiquorCtr`은 POJO인데 Redis를 고려해야할 필요가 있나 싶어서요...

Reply:
> 넵 근데 Redis에서 메모리로 가져온 데이터는 참조를 들고 있으면 언제든 재사용할 수 있는거 아닌가요? Redis에서 만료된 데이터가 WAS에서 계속 참조 하고 있으면 돌아다닌다는 것을 말하고 싶었습니다.

당연히 그러겠지만 요청이 끝난 후에도 어떻게 참조한 값을 계속 유지할 수 있는지 이해가 잘 되지 않네여..

> LiquorCtrTest는 도메인 단위 테스트인데 Redis를 고려하는게 이상하다는 점에서 시작한 코멘트였습니다. 아무리 생각해도 LiquorCtr은 POJO인데 Redis를 고려해야할 필요가 있나 싶어서요...

LiquorCtrTest, LiquorCtrRepositoryTest 모두 Redis를 고려하지 않고 있는데 혹시 어느 부분 말씀하신 건가여??

Reply:
## comment 1
> 당연히 그러겠지만 요청이 끝난 후에도 어떻게 참조한 값을 계속 유지할 수 있는지 이해가 잘 되지 않네여..

맞습니다. 억지로 상황을 만들어 봤습니다! 크게 신경은 안쓰셔도 될 것 같아요.

## comment 2

> LiquorCtrTest, LiquorCtrRepositoryTest 모두 Redis를 고려하지 않고 있는데 혹시 어느 부분 말씀하신 건가여??

최초에 코멘트를 단 이유는 클릭률을 구하는 곳에서 `overwrite()`를 호출해서 왜 호출했는지 물어본 것이었습니다. 그냥 LiquorCtr을 가져오거나, 생성해서 바로 구하면 되기 때문이죠.

제가 했던 질문은 다음과 같습니다.

> 뭔가 해당 given 절에서는 overwrite도 있어서 순수한 getCtr() 테스트가 아닌 것 같아요.

이때 다음과 같이 답변해주셨습니다.

> 첫 번째 코멘트 답변:
> 
> Redis 데이터가 만료됨 + 만료 이벤트 발생
> 만료 이벤트를 통해 만료된 데이터가 WAS 적재됨
> WAS 상에서 지연 쓰기 관련 만료 로직 처리 (RDB에 반영하는 로직)
> 요 세 단계는 이벤트 관련 로직이구요
> 
> 코멘트 남겨주신 코드는 그냥 redis에 적재되어 있던 데이터를 가져오는 코드입니다!

`overwrite()` 코드에 대해 물어봤는데 이벤트와 Redis에 적재되어있던 데이터라고 말씀해주셔서 순수한 도메인 테스트에서 Redis를 고려하고 있는건 아닐까? 하는 질문이었습니다.

다시 돌아와서,  클릭률을 구하는 `getCtr()`을 테스트하는 코드에 `overwrite()`를 작성하신 이유가 궁금합니다.

Reply:
> 최초에 코멘트를 단 이유는 클릭률을 구하는 곳에서 overwrite()를 호출해서 왜 호출했는지 물어본 것이었습니다. 그냥 LiquorCtr을 가져오거나, 생성해서 바로 구하면 되기 때문이죠.

앗 현재 코드는 바로 아래의 피드백이 반영되어 있는 상태입니다!

> 뭔가 해당 given 절에서는 overwrite도 있어서 순수한 getCtr() 테스트가 아닌 것 같아요.

> 위처럼 변경하고 overwrite 테스트도 추가하면 어떨까요?

이 피드백을 받기 전까지 제가 크게 놓치고 있어서..ㅠㅠㅠㅠㅠ
그래서 getCtr()에서 overwrite() 메서드 호출부를 제거했고, overwrite() 테스트도 추가한 상태입니다!!

Reply:
아하! ㅇㅅㅇ(역시 소통은 어렵다)